SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyRefCasts.swift

```diff
@@ -86,6 +88,15 @@ private extension UnaryInstruction {
       operand.set(to: replacement, context)
     }
 
+    if let ccb = self as? CheckedCastBranchInst {
+        // Make sure that loosing the formal source type is harmless (i.e the
```

Comment:
I don't understand this comment. I think what you want to check is the following:
if the lowered source operand type _is_ the formal type before the transformation we can assume that the type of the new operand is also a formal type.

Is this assumption always true?
Wouldn't it be better to just check if the new operand type `isLegalFormalType()`?

Reply:
Forgot to update the comment after updating the code. Have you seen my comment on this PR above?
 
Yes you are right: The comment  should say: if the lowered source operand type is the formal type before the transformation we can assume that no information was lost during type lowering.

Does isLegalFormalType check this condition?

It is not clear to me that this is why checked _cast_addr_br needs the formal source type:
https://github.com/swiftlang/swift/pull/67583

Reply:
> It is not clear to me that this is why checked _cast_addr_br needs the formal source type:

It's used in IRGen (`visitCheckedCastAddrBranchInst`). I assume the runtime type/cast machinery needs formal types rather than lowered types.

> Does isLegalFormalType check this condition?

I think if the new SILType is a formal type (which `isLegalFormalType` should check) it's fine to use it.

Reply:
Okay, let's use that.